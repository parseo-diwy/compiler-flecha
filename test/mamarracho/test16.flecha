
---- Print

def unsafePrintString s =
  case s
  | Nil       -> 0
  | Cons x xs -> (unsafePrintChar x; unsafePrintString xs)

def printList list =
  case list
  | Nil       -> 0
  | Cons x xs ->
    (print x;
     if null xs
      then 0
      else (unsafePrintString ", "; printList xs))

def print x =
  case x
  | Int       -> unsafePrintInt x
  | Char      -> unsafePrintChar x
  | False     -> unsafePrintString "False"
  | True      -> unsafePrintString "True"
  | Closure   -> unsafePrintString "<fun>"
  | Nil       -> unsafePrintString "[]"
  | Cons x xs -> (unsafePrintString "[";
                  printList (Cons x xs);
                  unsafePrintString "]")

def println x = print x; print '\n'

---- Listas

def null list =
  case list
  | Nil       -> True
  | Cons x xs -> False

def map f l =
  case l
  | Nil       -> Nil
  | Cons x xs -> Cons (f x) (map f xs)

---- Main

def main = 
  println (map (\ x -> 42)
               (Cons 1 (Cons 2 (Cons 3 Nil))));

  println (map (\ x -> Cons x (Cons x Nil))
               (Cons 1 (Cons 2 (Cons 3 Nil))))

