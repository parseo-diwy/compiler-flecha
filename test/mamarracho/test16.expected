[42, 42, 42]
[[1, 1], [2, 2], [3, 3]]
